### ISSUE TITLE
Virtual num pad is missing cancel button

### ISSUE THREAD

**Issue by Branikolog** [ihhub/fheroes2#9302]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

It seems like sometimes player changes his mind and wants just to skip number selection.

https://github.com/user-attachments/assets/a9a391ae-e6ae-48d2-b062-4c53030a29b0

Without cancel button and small available range of numbers it could be quite annoying to get back to the previous dialog.

### Describe the possible solution

I suggest adding "Cancel" button to this dialog for easier returning to the previous window.

**Comment by oleg-derevenetz** [comment]
This virtual numpad is just a variant of the virtual keyboard. Have you ever seen the "Cancel" button anywhere on devices with a virtual keyboard (Android, iOS)? :)

**Comment by Branikolog** [comment]
Hi, @oleg-derevenetz !

>Have you ever seen the "Cancel" button anywhere on devices with a virtual keyboard (Android, iOS)? :)

Well, there's always a device default cancel button in the bottom of the screen, that anyone can press to cancel keyboard. :)
I understand your point, but in a current implementation it is sometimes too difficult to return back to the previous dialog. If we're working with a tiny creature number (like - 1-3 monsters), we need to click precisely within these three buttons. If a player either had forgotten the initial number of available monsters, or trying to buy monsters having insufficient resources, he comes up with warning message, that he has to select specific numbers. He has to read the whole message to find out the range available or select numbers randomly trying to guess the number, that game will accept. By that time he obviously can decide not to select any number and just go back focusing on some other actions.
So there is no critical issue here, other than the current behavior is just a little unfriendly.

Virtual numpad could be cancelled with a click on the area outside of the window, but it lies out of the overall UI design and simply was not implemented before (requiring new logic for implementing such behaviour). So in my opinion having "cancel" button is a nice solution to return back without thinking what number I have to click to close this numpad keyboard.

If you think this enhancement is questionable, you are free to convert the issue to discussion. 🌞

**Comment by zenseii** [comment]
@Branikolog has a point here. In Android, virtual keyboards can be canceled by either the system's cancel button in the bottom home bar, or by clicking outside of the keyboard area. None of these are present in the fheroes2 UI, and thus the fheroes2 UI is inconsistent with users' expectations of virtual keyboards.

**Comment by oleg-derevenetz** [comment]
Hi @zenseii 

> In Android, virtual keyboards can be canceled by either the system's cancel button in the bottom home bar, or by clicking outside of the keyboard area.

Strictly speaking, they are not "canceled". Yes, they are removed from the screen, but everything that has already been entered using the keyboard in the corresponding input remains.

**Comment by zenseii** [comment]
@oleg-derevenetz. Yes, you're right. We can simply call the button something more correct, like Exit, Leave, Back etc..

**Comment by oleg-derevenetz** [comment]
> Yes, you're right. We can simply call the button something more correct, like Exit, Leave, Back etc..

So the existing OKAY button and this hypothetical Exit/Leave/Back button will in fact behave the same way? :)

**Comment by zenseii** [comment]
> So the existing OKAY button and this hypothetical Exit/Leave/Back button will in fact behave the same way? :)

According to the opening post, this would not be the case. The Okay button serves to check and go ahead with the entered input, the other button would ignore any input given and simply close the dialog.

**Comment by zenseii** [comment]
If we don't want to add more buttons to this dialog, then perhaps we could make the title of the dialog show the allowed input range. Ex. `Please input a value between 0-8`. After all, the issue described here arises when a user doesn't remember the allowed range and struggles to exit the virtual num pad.

**Comment by oleg-derevenetz** [comment]
> the other button would ignore any input given and simply close the dialog

But the system virtual keyboards don't work that way :) They do not magically return to the old state of the corresponding input after they are closed.

**Comment by zenseii** [comment]
> But the system virtual keyboards don't work that way :) They do not magically return to the old state of the corresponding input after they are closed.

Yep, the analogy with the system virtual keyboards was never perfect because our virtual numpad, which is admittedly what we should have discussed and not the fheroes2 virtual keyboard, is different in nature compared to the system virtual keyboards, in that the virtual numpad has a restriction in what it accepts as final input.
Meanwhile, a system virtual keyboard does not have this restriction and the user can input whatever they want, provided it is on the keyboard.

**Comment by oleg-derevenetz** [comment]
> Meanwhile, a system virtual keyboard does not have this restriction and the user can input whatever they want, provided it is on the keyboard.

Well, yes, but if we (hypothetically) used the system numpad for this, and not our own, then we would still do the check, it would just work in the "main" window, and not in the auxiliary numpad window. In either case, there would be no "magical" return to the previous value.

Technically, of course, we can add a "cancel" button to the virtual numpad. But there will definitely be questions like "why is the cancel button in the numpad, but not in the full virtual keyboard?" :)

**Comment by zenseii** [comment]
> Well, yes, but if we (hypothetically) used the system numpad for this, and not our own, then we would still do the check, it would just work in the "main" window, and not in the auxiliary numpad window. In either case, there would be no "magical" return to the previous value.

Yep, and this is just how Heroes 2 works originally and I see no reason to change where the check is done.

> Technically, of course, we can add a "cancel" button to the virtual numpad. But there will definitely be questions like "why is the cancel button in the numpad, but not in the full virtual keyboard?" :)

Yes, as I said above, this issue might as well be considered a lack of information issue, rather than a lack of a button issue:

> If we don't want to add more buttons to this dialog, then perhaps we could make the title of the dialog show the allowed input range. Ex. `Please input a value between 0-8`. After all, the issue described here arises when a user doesn't remember the allowed range and struggles to exit the virtual num pad.

**Comment by ihhub** [comment]
Let's go with displaying the allowed range of numbers. This is going to be much more user friendly than inventing the way of exit button.
